Добавить пайплайн очистки в CI (#8)

birthdaysgift opened the pull request:
После того как мы добавили линтеры в CI пайплайн с помощью GitHub Actions (#6), на вкладке [Actions](https://github.com/week-password/wisher-backend/actions) стали отображаться запущенные и завершенные раны нашего пайплайна. Так как мы используем ребейз-скваш флоу при работе с гитом, то очень часто будет возникать ситуация в которой ветка была запушена с хешами коммитов, например, "1" и "2", но потом ребейзнута, и хэши коммитов поменяются на "3" и "4". После этого у нас в нашем списке ранов на гитхабе останутся раны для коммитов "1" и "2". Но так как эти раны связаны с несуществующими на текующий момент коммитами, то нет смысла их хранить, т.к. они будут просто засорять список действительно полезных и нужных ранов для актуальных коммитов. Аналогичная ситуация будет происходить и при удалении веток - гитхаб будет хранить раны для коммитов в несуществующих ветках.

Удалять такие раны вручную и постоянно следить за этим довольно неудобно, поэтому в рамках настройки CI необходимо добавить воркфлоу, который будет автоматически удалять раны с гитхаба для несуществующих коммитов после ребейза ветки и после удаления ветки.

Rey-Lane commented on .github/workflows/cleanup.yml:
```diff
@@ -0,0 +1,66 @@
+      - name: Setup python
+        uses: actions/setup-python@v4
+        with:
+          python-version: "3.10"
```
Почему тут `Python 3.10`?

birthdaysgift replied:
По большому счету нет разницы какой тут пайтон, т.к. единственное его предназначение это выполнить cleanup-скрипт, а с этой задачей справится и 3.5)

Но вообще, если мы можем использовать последний пайтон, то почему бы его не использовать. Да и к тому же я уже обновил 3.10 -> 3.11 в предыдущей джобе, а здесь забыл

Rey-Lane replied:
Логика рассуждений понятна)
Просто, как говорится, `подоконник либо вровень, либо выпирает` 😅

Rey-Lane commented on envs/ci/scripts/github/cleanup/runs/rebased_branch.py:
```diff
@@ -0,0 +1,40 @@
+    cmd = subprocess.run(["git", "cherry", "origin/develop", GITHUB_REF_NAME], capture_output=True, encoding="utf-8")
+    for output_line in cmd.stdout.strip("\n").split("\n"):
+        commit_hash = output_line.split(" ")[-1]  # parse commit hash
+        yield commit_hash
```
Может это сделать как-то более безопасно?

```python
def get_branch_commits():
    repository = Github(GITHUB_TOKEN).get_repo(GITHUB_REPOSITORY)
    current_branch = repo.get_branch(GITHUB_REF_NAME)
    base_branch = repo.get_branch("develop")
    commits_diff = repo.compare(base_branch.commit.sha, current_branch.commit.sha)
    for commit in compare.commits:
        yield commit.sha
```
Не ручаюсь за работоспособность этого кода, примерный план

birthdaysgift replied:
Да, так лучше, и всё работает)

birthdaysgift committed:
8: Добавить пайплайн очистки в CI

После того как мы добавили линтеры в CI пайплайн с помощью GitHub Actions (#6), на вкладке [Actions](https://github.com/week-password/wisher-backend/actions) стали отображаться запущенные и завершенные раны нашего пайплайна. Так как мы используем ребейз-скваш флоу при работе с гитом, то очень часто будет возникать ситуация в которой ветка была запушена с хешами коммитов, например, "1" и "2", но потом ребейзнута, и хэши коммитов поменяются на "3" и "4". После этого у нас в нашем списке ранов на гитхабе останутся раны для коммитов "1" и "2". Но так как эти раны связаны с несуществующими на текующий момент коммитами, то нет смысла их хранить, т.к. они будут просто засорять список действительно полезных и нужных ранов для актуальных коммитов. Аналогичная ситуация будет происходить и при удалении веток - гитхаб будет хранить раны для коммитов в несуществующих ветках.

Удалять такие раны вручную и постоянно следить за этим довольно неудобно, поэтому в рамках настройки CI необходимо добавить воркфлоу, который будет автоматически удалять раны с гитхаба для несуществующих коммитов после ребейза ветки и после удаления ветки.

birthdaysgift commented:
Коммит, на который ссылается пулл реквест не существует, так как мы меняли соглашение по именованию коммитов и ребейзили `develop` ветку. Подробнее об этом можно почитать в [обсуджении](https://github.com/week-password/wisher-backend/discussions/3#discussioncomment-5640598).

Хэш нового коммита, содержащего точную копию изменений внесённых этим пулл-реквестом: 4a3402a
